### ISSUE TITLE
Parade example of careless Woo bloat, invalidating any chance for speedy page loads 

### ISSUE THREAD

**Issue by ppwebox** [woothemes/woocommerce#43879]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Once again Woo is forcing me to conduct research on improving page load speeds, after all of the usual tactics have already been utilized (page cache, object cache, compressing, minifying, deferring, dequeuing unused CSS/JS, etc., etc.). 

Measuring a Woo product archive page (small shop with maybe 50 products) with wp profile hook tells us the not so nice truth about Woocommerce:

WooCommerce Payments:
WC_Payments::init_rest_api()               | woocommerce-payments/includes/class-wc-payments. | 0.0139s
wcpay_jetpack_init()                       | woocommerce-payments/woocommerce-payments.php:75 | 0.023s
wcpay_init()                               | woocommerce-payments/woocommerce-payments.php:14 | 0.0999s
	Subtotal	0.13s

WooCommerce Blocks & Admin:
Automattic\WooCommerce\Admin\Features\Feat | woocommerce/src/Admin/Features/Features.php:132  | 0.024s
Automattic\WooCommerce\Blocks\BlockPattern | woocommerce/src/Blocks/BlockPatterns.php:92      | 0.055s
Automattic\WooCommerce\Internal\Admin\Prod | woocommerce/src/Internal/Admin/ProductReviews/Re | 0.0564s
Automattic\WooCommerce\Blocks\BlockTypesCo | woocommerce/src/Blocks/BlockTypesController.php: | 0.0893s
Automattic\WooCommerce\Admin\API\Init->res | woocommerce/src/Admin/API/Init.php:60            | 0.2374s
	Subtotal	0.46s

WooCommerce:
WC_Cart_Session->get_cart_from_session()   | woocommerce/includes/class-wc-cart-session.php:9 | 0.0055s
WC_Post_Types::register_taxonomies()       | woocommerce/includes/class-wc-post-types.php:39  | 0.0153s
wc_product_post_class()                    | woocommerce/includes/wc-template-functions.php:5 | 0.0277s
wc_setup_product_data()                    | woocommerce/includes/wc-template-functions.php:1 | 0.043s
Automattic\WooCommerce\Packages::on_init() | woocommerce/src/Packages.php:62                  | 0.054s
woocommerce_simple_add_to_cart()           | woocommerce/includes/wc-template-functions.php:1 | 0.0601s
wc_product_post_type_link()                | woocommerce/includes/wc-product-functions.php:21 | 0.0762s
WooCommerce->init()                        | woocommerce/includes/class-woocommerce.php:697   | 0.0872s
function(){}                               | woocommerce/src/StoreApi/StoreApi.php:25         | 0.0887s
Automattic\WooCommerce\RestApi\Server->reg | woocommerce/includes/rest-api/Server.php:39      | 0.3249s
	Subtotal	0.70s
	
	Total		1.37s (for Woocommerce related processing alone on a SMALL shop page)

Folks, it is getting worse and worse with the bloated Woo code by the day. And can anybody tell me why so many Woo functions and scripts are called/excuted on EVERY page? A website/shop's pages are usually 60-70% product related (requiring Woo) and 30-40% normal pages NOT requiring Woo. 

Next improvement: why is the heavy WooCommerce Admin stuff loaded on the frontend at all? Ditto for WooCommerce Blocks - may be 50-60% of all sites worldwide use a page builder and not WP blocks. Yet the Woo devs blindly, naively, or ignorantly just load the entire Woo enchillada anytime and everywhere. Who cares, right? We all are gonna die anyway at one point!  

At least give the site owners the option to turn off this nonsense. Poor page load time cost money as you loose customers, have fewer sales, achieve poorer search result rankings, etc. In addition, such careless coding produces a higher energy foot print (+CPU, hence more power consumption) each day if multiplied by the millions of WP/Woo web sites. 

In my opinion, we are approaching soon a turning point, where you seriously have to consider whether utilizing Woocommerce for eCommerce tasks makes any sense.  

Thank you for your time, gentlemen.

### Expected behavior

Significantly faster page load times, such as expected by Google for example.

### Actual behavior

Slow page load times due to Woocommerce bloat.

### Steps to reproduce

1. wp profile hook --all --orderby=time --spotlight --url=yoursite-with-woo-installed
2. Analyze timings
3. Be frustrated about Woo

### WordPress Environment

See wp hook results in the bug description.

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by mikejolley** [comment]
Hi there,

First off, thanks for taking the time to document your findings. There have been many projects merging into the monorepo of late so there is likely lots of room for improvement. I will address your comments first and then flag some of the individual results for team feedback.

> Can anybody tell me why so many Woo functions and scripts are called/excuted on EVERY page? A website/shop's pages are usually 60-70% product related (requiring Woo) and 30-40% normal pages NOT requiring Woo.

Generalising, its likely because functions need to be readily available for anything on the page that may want to consume them. During init its unlikely that any plugin could predict accurately which components will be needed during page load, and some things need to be called early (such as sessions/cookies etc).

While its true not all pages are "store" pages, its quite common to have WooCommerce elements site wide. Mini cart, product displays as some examples. Therefore certain components are loaded sitewide.

> Next improvement: why is the heavy WooCommerce Admin stuff loaded on the frontend at all? 

Thats a good question and something that needs to be addressed or documented.

> Ditto for WooCommerce Blocks - may be 50-60% of all sites worldwide use a page builder and not WP blocks. Yet the Woo devs blindly, naively, or ignorantly just load the entire Woo enchillada anytime and everywhere. 

As documented here:

- https://developer.wordpress.org/reference/functions/register_block_pattern/
- https://developer.wordpress.org/reference/functions/register_block_type/

Block types and patterns must be registered during `init`. There may be room for improvement in terms of caching or optimising methods which register blocks and patterns, but they need to be registered during `init` regardless.

Also note, the block based checkout _is_ default now. Shortcode checkout will be maintained but won't see any major improvements. So adapting to the block based experience will be essential. 

— 

## WooCommerce Payments

> WC_Payments::init_rest_api() | woocommerce-payments/includes/class-wc-payments. | 0.0139s
> wcpay_jetpack_init() | woocommerce-payments/woocommerce-payments.php:75 | 0.023s
> wcpay_init() | woocommerce-payments/woocommerce-payments.php:14 | 0.0999s

cc @woocommerce/moltres @woocommerce/fractal 

Could someone access these methods and ensure they are needed on frontend requests?

## WooCommerce Admin

> Automattic\WooCommerce\Admin\Features\Feat | woocommerce/src/Admin/Features/Features.php:132 | 0.024s
> Automattic\WooCommerce\Internal\Admin\Prod | woocommerce/src/Internal/Admin/ProductReviews/Re | 0.0564s
> Automattic\WooCommerce\Admin\API\Init->res | woocommerce/src/Admin/API/Init.php:60 | 0.2374s

cc @woocommerce/mothra 

## WooCommerce Blocks

> Automattic\WooCommerce\Blocks\BlockPattern | woocommerce/src/Blocks/BlockPatterns.php:92 | 0.055s
> Automattic\WooCommerce\Blocks\BlockTypesCo | woocommerce/src/Blocks/BlockTypesController.php: | 0.0893s

@woocommerce/rubik @woocommerce/woo-fse 

The registration of patterns may be a bit heavy because it:

- scans directories
- loads meta data from all files
- registers patterns

Could this be cached or could the patterns be defined manually to avoid loading from the filesystem?

`BlockTypesController` doesn't too anything too heavy but it does register assets and do all hydration on load. Perhaps this could be deferred to improve performance? 

cc @opr @senadir 

## WooCommerce Core

> WC_Cart_Session->get_cart_from_session() | woocommerce/includes/class-wc-cart-session.php:9 | 0.0055s
> WC_Post_Types::register_taxonomies() | woocommerce/includes/class-wc-post-types.php:39 | 0.0153s
> wc_product_post_class() | woocommerce/includes/wc-template-functions.php:5 | 0.0277s
> wc_setup_product_data() | woocommerce/includes/wc-template-functions.php:1 | 0.043s
> Automattic\WooCommerce\Packages::on_init() | woocommerce/src/Packages.php:62 | 0.054s
> woocommerce_simple_add_to_cart() | woocommerce/includes/wc-template-functions.php:1 | 0.0601s
> wc_product_post_type_link() | woocommerce/includes/wc-product-functions.php:21 | 0.0762s
> WooCommerce->init() | woocommerce/includes/class-woocommerce.php:697 | 0.0872s
> function(){} | woocommerce/src/StoreApi/StoreApi.php:25 | 0.0887s
> Automattic\WooCommerce\RestApi\Server->reg | woocommerce/includes/rest-api/Server.php:39 | 0.3249s

@woocommerce/proton For awareness. 

Do we need to load the Rest API loaded on the frontend? It's init here:

https://github.com/woocommerce/woocommerce/blob/1a4ebebf6a0157ff4882528453e7a53b7177ed9c/plugins/woocommerce/includes/class-woocommerce.php#L433-L438

Would it be sensible to wrap with `is_rest_api_request` or are there other cases we need to account for? If so, could we initialise it some other way to prevent it loading unneccessarily?

cc @nerrad

**Comment by ppwebox** [comment]
Thanks for igniting action, Mike!

These hosting guys experienced a CPU usage spike of plus 30% across their servers after a Woocommerce update to 7.2.3 with blocks: [https://onpointplugins.com/disabling-woocommerce-blocks/] 
And then quickly wrote a script (installed in /wp-content/mu-plugins/) to dequeue the Woo Blocks package for good: 

#<?php
use Automattic\WooCommerce\Packages;
add_action( 'plugins_loaded', function() {
    if ( ! class_exists( Packages::class ) ) {
		return;
	}
    // Dependencies are require for Woo admin screens.
	if ( ! empty( $_GET['page'] ) && is_admin() && str_starts_with( $_GET['page'], 'wc-' ) ) {
		return;
	}
	$override = new class() extends Packages {
		public function __construct() {
		}
		public function remove_woocommerce_blocks() : void {
			unset( static::$packages['woocommerce-blocks'] );
		}
	};
	$override->remove_woocommerce_blocks();
}, 6 );

Using this script will not affect a customized checkout, such as when built with Elementor or using the shortcode, right?

**Comment by ppwebox** [comment]
> While its true not all pages are "store" pages, its quite common to have WooCommerce elements site wide. 
> Mini cart, product displays as some examples. Therefore certain components are loaded sitewide.

We should have options here, since usually that's just the Mini cart. There must be a better option than just fully loading Woo. I for one use two slightly different headers for Woo and non-Woo pages (Woo dequeued by script) - with mini cart and without but with a link to the cart instead. 

One problem I noted with completely dequeuing Woo on certain pages: payment on checkout via PayPal is successful but the backend sync does not complete - order status stays at Pending instead completed.

**Comment by ppwebox** [comment]
The most pressing performance issue with 0.32 delay is Woocommerce's Rest Api Server loading on the frontend. This on a home page that does not even utilize Woo components!

**Comment by lkraav** [comment]
Yep, glad someone else is noticing these things cc @glagonikas 

I'm happy Woo team pushed a PR for https://github.com/woocommerce/woocommerce/issues/26630, at least it's possible now for advanced shops to control cart subsystem loading.

Blocks-based frontend React checkout mini-app future is a whole another story altogether, not even sure yet what we're going to do with that, should we stay on WP/WC.

See also https://github.com/woocommerce/woocommerce-blocks/discussions/8166

**Comment by vbelolapotkov** [comment]
@ppwebox @mikejolley thanks for the investigation and your report. I've taken insights for WooPayments and moved to its own repo: https://github.com/Automattic/woocommerce-payments/issues/8090

Feel free to follow the issue for monitoring progress. Unfortunately, it's hard to say at the moment when we'll be able to fix it.

**Comment by lanej0** [comment]
Going to send this one over Proton's way because it looks like it hasn't been picked up by them yet, and there's some outstanding questions there.

**Comment by vedanshujain** [comment]
hey @ppwebox (and others following the issue), note that the slowness you are experiencing is coming because of two main issues:

1. We are incorrectly loading REST API on any page where there is a WC block. This causes slowness because loading API itself is slow (see next point) which can be seen in the frontend requests.

2. REST API is slow because it loads all the controllers that are available on the site, whether or not they are needed to serve the request. No particular controller is super slow, it's just the sheer volume of code that we load in this case is quite high.


We have merged #44080 that should address point 1. You should be able to verify this by something as simple as a curl benchmarking. In my testing, I am seeing performance improvements of up to 2-3s on front page load on a strained entry level server. If you would like to test I would also suggest adding this small plugin somewhere (don't do in production as it will log stack trace) and inspecting browser log to verify that API is not being loaded by another plugin. Another caveat is that, this patch (#44080) won't show any improvement when running `wp profile` because that itself loads REST API which will defeat the improvements from this patch.

```php
<?php
/**
 * Plugin Name: API Profiler
 */

add_filter( 'rest_api_init', function () {
	if (  WC()->is_rest_api_request() ) {
		return;
	}
	$e = new \Exception;
	js_console_log( array(
		'message' => 'REST API loaded',
		'stack' => $e->getTraceAsString(),
	));

} );

function js_console_log( $data ) {
	?>
	<script>
		console.log( <?php echo json_encode( $data ); ?> );
	</script>
	<?php
}
```

Additionally, we have opened #44539 to address loading everything on REST API. We are hoping to address it soon as well, and once that's merged you would be able to see improvements in `wp profile`, as well as in general on all REST API requests.

I am closing this issue as I am planning to address further improvement in #44539. Happy to re-open, discuss and add patches to the issue though if there are additional suggestions! Also note that we are doing a holistic performance effort as part of this project: https://github.com/orgs/woocommerce/projects/220.

**Comment by rpldom** [comment]
This one are slowest functions in 'init' hook 

<img width="1829" height="262" alt="Image" src="https://github.com/user-attachments/assets/f244fe8c-118b-4bf7-98db-06bd40ca2302" />

After latest performance patches - still slow
